### PR TITLE
Levinska_fixedCurrency_EditAccountTests

### PR DESCRIPTION
**Pull request by sofiialevinska**
method getCurrencyButton() simplified; getMacBookCost locator and method deleted; Searching of product containers and checking if their prices is correct moved from Test to BL; ChangeEmail, ChangePassword, ChangeName tests and LoginUsingValidData_Logout class moved into EditAccountTest class; additional verification of password added.

### CODE REVIEW

**Comment by sofiialevinska** [comment]
@adomanich ми випадково змерджили минулий пул реквест (https://github.com/sofiialevinska/Report1947/pull/2) з мастер. 

такого більше не має повторитись - застосувала правило 3х рев'юверів перед мерджем до гілки master (бо правила були, але не були застосовані до жодної з гілок).

тому створила новий пул реквест - цей. В нього внесені зміни по твоїм коментарям. на коментарі відповіла у тому пул реквесті і там є питання.

File: src/main/java/com/opencart/steps/HeaderPageBL.java [hunk 1]
```diff
@@ -68,18 +66,29 @@ public HeaderPageBL clickOnCurrencyButton(String currencyName) {
...
      */
 
-    public HeaderPageBL verifyMacBookCost(ProductContainer container, String currencyName) {
+    public void verifyAllProductsPrices(String currencyName) {
+        List<ProductContainer> productContainers = new SearchResultPage().getProductContainers();
+        for (ProductContainer container
+                : productContainers) {
+            verifyProductCost(container, currencyName);
+        }
+    }
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/HeaderPageBL.java]: I think it would be better to move this logic to MainPageBL?

File: src/main/java/com/opencart/steps/HeaderPageBL.java [hunk 2]
```diff
@@ -68,18 +66,29 @@ public HeaderPageBL clickOnCurrencyButton(String currencyName) {
...
 
+    /**
+     * Method checks if MacBook Cost in specified currency is correct.
+     *
+     * @param container    is Container for Products where price needs to be checked;
+     * @param currencyName is name of Currency that needs to be checked.
+     */
+    private void verifyProductCost(ProductContainer container, String currencyName) {
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/HeaderPageBL.java]: How do you think is it a good approach to hardcode the real values of different currencies?
Imagine the situation when somebody goes to the Admin page and change it, your test will fail then.

So, I would recommend you do the next:
1. Go to the admin page and get real values from existing currencies, then add also add a new one by the autotest. 
2. Perform tests with different currencies including the newly created one.
3. You need to remove already created currency after your test finished.

Comment (sofiialevinska) [reply src/main/java/com/opencart/steps/HeaderPageBL.java]: okay. 

Should I also go to AdminPage (screenshot) and read actual Product Prices from there?
and then compare them with existing ones on the website?

  Screen Shot 2020-11-20 at 13 09 15

**Comment by sofiialevinska** [comment]
@adomanich I have changed code according to comments.

1. add new currency through admin page is added
2. add reading Currencies values from Admin Page and test for setting new Currency through Admin Page

File: src/main/java/com/opencart/datamodel/AdminLoginModel.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package com.opencart.datamodel;
+
+public class AdminLoginModel {
+
+    private String username = "admin";
+    private String password = "admin";
+
+    private String currencyName = "UAH";
+    private String currencyValue = "28.22";
```
Comment (adomanich) [inline src/main/java/com/opencart/datamodel/AdminLoginModel.java]: How those fields relate to AdminLoginModel?

File: src/main/java/com/opencart/pages/AdminDashboard.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import com.opencart.containers.CurrencyContainerAdmin;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class AdminDashboard extends BasePage {
```
Comment (adomanich) [inline src/main/java/com/opencart/pages/AdminDashboard.java]: I think it is better to rename it to CurrencyAdminPage.
AdminDashboard page is a default page for "admin" panel, the same as Main page for "store".
So, that means that AdminDashboard page opens firstly when your login to "admin" panel.
Then you need to navigate to AdminCurrency page from there. For that, describe the left-navigation panel elements and business logic within separate classes, then just aggregate that logic to AdminDashboard (as we have done with HeaderPage).

File: src/main/java/com/opencart/steps/AdminDashboardBL.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        adminDashboard.getInputValue().sendKeys(currencyValue);
+    }
+
+    private void clickOnSaveCurrencyButton () {
+        adminDashboard.getSaveNewCurrency().click();
+    }
+
+//    public AdminDashboardBL deleteNewCurrency () {
+//        return this;
+//    }
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/AdminDashboardBL.java]: Please, remove commented code if you do not use it.

File: src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                .getHeaderPageBL()
+                .clickOnChangeCurrencyButton()
+                .clickOnCurrencyButton(data);
+        mainPageBL.verifyAllProductsPrices(data);
+    }
+
+    @AfterClass
+    public void closeBrowser () {
+        DriverRepository.closeBrowser ();
+    }
```
Comment (adomanich) [inline src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java]: We already have defined AfterClass() into BaseTest, do we need to create another one here?

File: src/main/java/com/opencart/steps/AdminDashboardBL.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        adminDashboard = new AdminDashboard();
+    }
+
+    public AdminDashboardBL clickOnSidebarButton (String buttonName) {
+        adminDashboard.getButton(buttonName).click();
+        return this;
+    }
+
+    public void getCurrencyValues() {
+        new CurrencyModel();
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/AdminDashboardBL.java]: You can return Map<String, String> from that method, and put logic from CurrencyModel class to there instead of creating a separate class <CurrencyModel> for one single method.

File: src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    @BeforeClass
+    public void getCurrenciesValuesFromAdminPage() {
+        new Navigation().navigateToURrl(ADMIN_BASE_URL.getValue());
+        new AdminLoginPageBL().adminLogin();
+        new AdminDashboardBL()
+                .clickOnSidebarButton("System")
+                .clickOnSidebarButton("Localisation")
+                .clickOnSidebarButton("Currencies")
+                .getCurrencyValues ();
```
Comment (adomanich) [inline src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java]: According to changes in that method which I describe above, you will get the following: 
Map<String, String> currencies = adminDashBoard.getCurrencyValues ();

And, also, you do not need to use a data-provider here, because you are hardcoding currencies there. It is not right, because you can not be sure that exactly the same values as into the data-provided will be returned by getCurrencyValues () method.

To sum up:
1. Remove data-provider
2. Update verifyAllProductsPrices() method, it should take Map<String, String> currency values got from admin page.

**Review by adomanich — CHANGES_REQUESTED** [review]

File: src/main/java/com/opencart/datamodel/AdminLoginModel.java (hunk 3 above)
Comment (sofiialevinska) [reply src/main/java/com/opencart/datamodel/AdminLoginModel.java]: I will move them to CurrencyModel

File: src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java (hunk 6 above)
Comment (sofiialevinska) [reply src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java:51]: It's for closing the Browser that opens AdminPage.

File: src/main/java/com/opencart/steps/AdminDashboardBL.java (hunk 7 above)
Comment (sofiialevinska) [reply src/main/java/com/opencart/steps/AdminDashboardBL.java]: I also use CurrencyModel class to check the checkbox (before deleting new Currency).  
Screen Shot 2020-11-22 at 12 39 14

Comment (adomanich) [reply src/main/java/com/opencart/steps/AdminDashboardBL.java]: You can add take currency name you want to delete as a method parameter, without using CurrencyModel class.

File: src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java (hunk 8 above)
Comment (adomanich) [reply src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java]: You have not finished make changes according to this comment yet

Comment (sofiialevinska) [reply src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java]: I suggest living dataProvider for USD, EUR, GBP currencies because they are standard for opencart and already implemented. Or is it better to write separate tests for USD, EUR, GBP currencies? Or to read from admin page names of existing currencies and then check those currencies?

File: src/main/java/com/opencart/steps/MainPageBL.java [hunk 9]
```diff
@@ -1,9 +1,66 @@
...
     }
+
+    /**
+     * Method finds all ProductContainers on the Page and checks
+     * if their Product Cost in specified currency is correct.
+     *
+     * @param currencyName is name of Currency that needs to be checked.
+     */
+
+    public void verifyAllProductsPrices(Map<String, String> currencies, String currencyName) {
```
Comment (adomanich) [inline src/main/java/com/opencart/steps/MainPageBL.java]: I believe you can get the currency name from the map. Am I wrong?

File: src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                .clickOnLeftNavigationPanelButton("Localisation")
+                .clickOnLeftNavigationPanelButton("Currencies");
+        currencies = new AdminCurrencyPageBL().getCurrencyValues();
+    }
+
+    @DataProvider(name = "data-provider-currency")
+    public Object[][] dataProviderMethod() {
+        return new Object[][]{{CurrencyName.EUR.getValue()}, {CurrencyName.USD.getValue()},
+                {CurrencyName.GBP.getValue()}};
+    }
```
Comment (adomanich) [inline src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java]: You do not need to use this Data-Provider, you already have all necessary data into currencies map

**Review by adomanich — CHANGES_REQUESTED** [review]

File: src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java (hunk 10 above)
Comment (sofiialevinska) [reply src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java]: aaa.. I think I got it.
I delete the data provider and hardcoded currency names and made taking currency names from Map that is created on Admin Page.

File: src/main/java/com/opencart/steps/MainPageBL.java (hunk 9 above)
Comment (sofiialevinska) [reply src/main/java/com/opencart/steps/MainPageBL.java]: yes) now I see it.

File: src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+                .clickOnLeftNavigationPanelButton("Currencies");
+        currencies = new AdminCurrencyPageBL().getCurrencyValues();
+    }
+
+    @Test(description = "Test checks if the cost of all products on the website's main page is correct " +
+            "for currencies that are presented on admin page")
+    public void ChangeCurrencyTest() {
+        new Navigation().navigateToURrl(BASE_URL.getValue());
+        MainPageBL mainPageBL = new MainPageBL();
+        currencies.forEach((name, value) -> {
```
Comment (adomanich) [inline src/test/java/com/opencart/tests/currency/ChangeCurrencyTest.java:37]: Please, perform iteration via map into verifyAllProductsPrices() method,  instead of doing it within the test
